[C++ for OpenCL] Document further deviation from OpenCL C (#262)
KhronosGroup/OpenCL-Docs: cxx4opencl/diff2openclc.txt

=== Difference to OpenCL C

{cpp} for OpenCL provides backwards compatibility with OpenCL C
for the majority of features. However, there are a number of
exceptions that are described in this section. Some of them come
from the nature of C++ but others are due to improved behaviour

> of a kernel language.

Reviewer: This reads a bit strange, do you mean to say improvements that have been added just for C++ for OpenCL?

Author: I changed it now as follows:

>  Some of them come
> -from the nature of C++ but others are due to improved behaviour
> -of a kernel language.
> +from the nature of C++ but others are due to improvements in OpenCL
> +features. Most of such improvements do not invalidate old code,
> +but simply provide extra functionality.

==== C++ related differences

{cpp} for OpenCL is a different language derived from {cpp} and

> therefore it inherits its fundamental design principles. Hence

Reviewer: its is ambiguous. I think it would be better to say c++s fundamental design principles

{cpp} for OpenCL deviates from OpenCL C in the same areas where
{cpp} deviates from C. The main reasons for that is to provide

> more helpful language to the developers or compilation tools.

Reviewer: a more helpful

=== Difference to OpenCL C

{cpp} for OpenCL provides backwards compatibility with OpenCL C
for the majority of features. However, there are a number of
exceptions that are described in this section. Some of them come

> from the nature of C++ but others are due to improvements in OpenCL

Reviewer 2: ```suggestion
from the nature of {cpp} but others are due to improvements in OpenCL
```

features. Most of such improvements do not invalidate old code,
but simply provide extra functionality.

==== C++ related differences

> {cpp} for OpenCL is a different language derived from {cpp} and

Reviewer 3: Different from what?  I'm assuming it's meant to mean OpenCL C.

Author: Yep ok good point. I will modify this. Thanks!

Author: Updated!

therefore it inherits {cpp}'s fundamental design principles. Hence
{cpp} for OpenCL deviates from OpenCL C in the same areas where

> {cpp} deviates from C. The main reasons for that is to provide a

Reviewer 2: ```suggestion
{cpp} deviates from C. This results in a
```

> more helpful language to the developers or compilation tools.

Reviewer 2: ```suggestion
more helpful language for developers and avoids substantially
increasing complexity of compilation tools.
```

   // ...
----------

==== OpenCL specific difference

> This section described where {cpp} for OpenCL differs from OpenCL C

Reviewer: described->describes

in OpenCL specific behavior.

===== Variadic macros

> {cpp} for OpenCL eliminates restriction on variadic macro from

Reviewer: macro->macros

> section 6.9.e of OpenCL C. Variadic macros can be used regularly

Reviewer: regularly->normally perhaps? regularly sounds a bit strange in this context

as per {cpp}17 [cpp.replace].

===== Predefined macros

> Predefined version macro `__OPENCL_C_VERSION__` and `NULL` described

Reviewer: macro->macros

> Predefined version macro `__OPENCL_C_VERSION__` and `NULL` described

Reviewer: 'NULL', described

> in OpenCL C 2.0 section 6.10 are not supported.

Reviewer: 6.10, are

> There is a macro `__OPENCL_CPP_VERSION__` available that is set to

Reviewer: Maybe these should be presented as a list rather than a paragraph?

value `100`. For convenience there is also `__CL_CPP_VERSION_1_0__`
that is defined to the same value `100` and can be used instead of a
literal.

===== Atomic operations

{cpp} for OpenCL relaxes restriction from OpenCL C 2.0 section 6.13.11 to

> atomic types by allowing to modify variables of such types using builtin

Reviewer: Perhaps this could be worded as "to allow them to be used by builtin operators"

operators, and not only using builtin function.

The behavior of operators on atomic types follows the logic from {cpp}17
sections [atomics.types.int][atomics.types.pointer][atomics.types.float].

[source,c]
----------
atomic_int acnt;
acnt++; // equivalent to atomic_fetch_add_explicit(&acnt, 1);
----------

===== Use of Clang Blocks

Clang Blocks that are defined by the Objective-C language are not supported
and their use can be replaced by lambdas ({cpp}17 `[expr.prim.lambda]`).

> This also implies that builtin functions using blocks, such as `enqueue_kernel`,

Reviewer: Seems slightly unusual to start a paragraph with This also implies. Perhaps say something like "The above implies ... "

   // ...
----------

==== OpenCL specific difference

This section describes where {cpp} for OpenCL differs from OpenCL C
in OpenCL specific behavior.

===== Variadic macros

> {cpp} for OpenCL eliminates restriction on variadic macros from `OpenCL C v2.0 s6.9.e`.

Reviewer 2: ```suggestion
{cpp} for OpenCL eliminates the restriction on variadic macros from
`OpenCL C v2.0 s6.9.e`.
```

Variadic macros can be used normally as per {cpp}17 `[cpp.replace]`.

===== Predefined macros

> Predefined version macro `__OPENCL_C_VERSION__` and macro `NULL`,

Reviewer 2: ```suggestion
The predefined macros `__OPENCL_C_VERSION__` and `NULL`,
```

> described in `OpenCL C v2.0 s6.10`, are not supported.

Reviewer 3: I'm aware that this PR is just documenting existing behaviour rather than a proposed design, so this comment can be resolved without any change, but:

Why does `C++ for OpenCL` exclude `NULL` when it is generally backwards compatible with OpenCL C 2.0?  I'm of course aware that `nullptr` is more idiomatic in modern C++, but even there [NULL is still reachable through the C headers](https://en.cppreference.com/w/cpp/types/NULL).

It just seems so easy to have something like:

```c
#define NULL nullptr
```

Author: Yep, I agree. One danger to that is an old definition of NULL that hasn't been the same so aliasing `NULL` to `nullptr` still doesn't provide 100% compatibility.

Of course, developers can do this trick manually if it suits their use case.

C++ for OpenCL can't achieve full backwards compatibility to OpenCL C, because it is a dialect of C++ rather than C. And in C++ there were fundamental changes that diverged from C often for good reasons I would say.

Reviewer 3: It still seems a surprising omission to me, but it's not the point of this PR so I'll resolve this comment.

The following new predefined macros are added in {cpp} for OpenCL:

 * `__OPENCL_CPP_VERSION__` set to value `100`.
 * `__CL_CPP_VERSION_1_0__` also set to `100` and can be used for
   convenience instead of a literal.

===== Atomic operations

{cpp} for OpenCL relaxes restriction from `OpenCL C v2.0 s6.13.11` to
atomic types allowing them to be used by builtin operators, and not
only by builtin functions.

> The behavior of operators on atomic types follows the logic from {cpp}17

Reviewer 2: ```suggestion
Operators on atomic types behave as described in {cpp}17
```

sections `[atomics.types.int]` `[atomics.types.pointer]`
`[atomics.types.float]`.

[source,c]
----------
atomic_int acnt;

> acnt++; // equivalent to atomic_fetch_add_explicit(&acnt, 1);

Reviewer 3: ```suggestion
acnt++; // equivalent to atomic_fetch_add(&acnt, 1);
```

The explicit form requires at least an order, or an order and a scope.

----------

===== Use of Clang Blocks

Clang Blocks that are defined by the Objective-C language are not supported
and their use can be replaced by lambdas ({cpp}17 `[expr.prim.lambda]`).

> The above implies that builtin functions using blocks, such as `enqueue_kernel`,

Reviewer 3: To me the original wording was stronger, I assume that those builtins really don't work.

Author: Yes, they are not going to be accepted in C++ for OpenCL mode.
